File: src/main/java/com/vaadin/demo/component/combobox/ComboBoxCustomEntry1.java
package com.vaadin.demo.component.combobox;

import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("combo-box-custom-entry-1")
public class ComboBoxCustomEntry1 extends Div {

  public ComboBoxCustomEntry1() {
    // tag::snippet[]
    ComboBox<String> comboBox = new ComboBox<>("Browser");
    comboBox.setAllowCustomValue(true);
    add(comboBox);
    // end::snippet[]
    comboBox.setItems("Chrome", "Edge", "Firefox", "Safari");
    comboBox.setHelperText("Select or type a browser");
  }

  public static class Exporter extends DemoExporter<ComboBoxCustomEntry1> { // hidden-source-line
  } // hidden-source-line
}

File: frontend/demo/component/combobox/combo-box-custom-entry-1.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/combo-box';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('combo-box-custom-entry-1')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private items = ['Chrome', 'Edge', 'Firefox', 'Safari'];

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-combo-box
        allow-custom-value
        label="Browser"
        helper-text="Select or type a browser"
        .items="${this.items}"
      ></vaadin-combo-box>
      <!-- end::snippet[] -->
    `;
  }
}
